ISSUER_PROTOCOL.md

```diff
@@ -0,0 +1,345 @@
+
+```
+Issue:
+  issued = min(count, toIssue)
+  xb,yb = secretKey.x1, secretKey.y1
+  if privateMetadata == 0:
+    xb,yb = secretKey.x0, secretKey.y0
+  signed = []
+  T,S,W,Ws = [], [], [], []
+  for i in 0..issued:
+    T' = nonces[i]
+    s ←$ {0, 1}λ
+    S' = Hs(T', s)
+    Wp = xb*T' + yb*S'
+    Wsp = secretKey.xs*T' + secretKey.ys*S'
+    T,S,W,Ws += T', S', Wp, Wsp
+    signed += SignedNonce{s, Wp, Wsp}
+  proof = DLEQbatched.P((publicKey,T,S,W,Ws),(secretKey.xs, secretKey.ys, xb, yb))
+  return (issued, keyID, signed, proof)
+
+DLEQbatched.P((X,T,S,W,Ws),(xs, ys, xb, yb)):
+  e0,...,em = Hc(X,T,S,W,Ws)
+  Tt = Sum(ej*Tj) // over j from 0 to m
+  St = Sum(ej*Sj) // over j from 0 to m
+  Wt = Sum(ej*Wj) // over j from 0 to m
+  Wst = Sum(ej*Wsj) // over j from 0 to m
+  dleqProof = DLEQ2.P((X,T,S,Ws), (xs,ys)) // cs,us,vs
+  dleqorProof = DLEQOR2.P((X,T,S,W), (xb,yb)) // c0,c1,u0,u1,v0,v1
+  return dleqProof + dleqorProof
+```
+
+Input Serialization:
+
+The Trust Token Issuance Request contains an `IssueRequest` structure defined below.
+
+```
+struct {
+  uint16 count;
+  ECPoint nonces[count];
+} IssueRequest;
+```
+
+Output Serialization:
+
+The Trust Token Issuance Response contains an `IssueResponse` structure defined below.
+
+```
+struct {
+  opaque s<Nn>; // big-endian bytestring
+  opaque Wp<1..2^16-1>; // X9.62 Uncompressed Wp point.
+  opaque Wsp<1..2^16-1>; // X9.62 Uncompressed Wsp point.
+} SignedNonce;
+
+struct {
+  Scalar cs;
+  Scalar us;
+  Scalar vs;
+  Scalar c0;
+  Scalar c1;
+  Scalar u0;
+  Scalar u1;
+  Scalar v0;
+  Scalar v1;
+} DLEQProof;
+
+struct {
+  uint16 issued;
+  uint32 key_id = keyID;
+  SignedNonce signed[issued];
+  opaque proof<1..2^16-1>; // Length-prefixed form of DLEQProof.
+} IssueResponse;
+```
+
+### Redeem Function
+
+The _Redeem_ function corresponds to the **AT.VerValid** and **AT.ReadBit** stage of the PMBTokens protocol.
+
+Inputs:
+
+*   token (The trust token to redeem)
+*   client\_data (the client data sent as part of the redemption request to include in the SRR)
+*   redemptionTime (the redemption time from the client)
+*   secretKey (the secret key that should be used to sign this request, determined by the public metadata)
+*   lifetime (lifetime for the SRR)
+*   keys (dictionary from known key IDs to secret/public keys)
+*   srrKey (SRRSigningKey)
+
+Outputs:
+
+*   srr (the Signed Redemption Record, a binding of the client data and metadata to this redemption)
+*   signature (the signature over the SRR)
+
+```
+Redeem:
+  secretKey, publicKey = keys[token.key_id]
+  T = Ht(token.nonce)
+  if token.Ws != secretKey.xs*T+secretKey.ys*token.S:
+    return 0 // Invalid validity verification.
+  privateMetadata = 0
+  if token.W == secretKey.x0*T+secretKey.y0*token.S:
+    privateMetadata = 0
+  elseif token.W == secretKey.x1*T+secretKey.y1*token.S:
+    privateMetadata = 1
+  else:
+    return 0 // Internal Error
```

Comment:
This part is somewhat timing-sensitive. Is it worth pointing that out, and perhaps writing it in that way? (I.e. do both operations and then I guess write pseudocode for the equality booleans, with a comment that it's timing-sensitive.)

Reply:
Done.